tendermint/src/block/id.rs

```diff
@@ -34,12 +40,92 @@ pub struct Id {
     /// way to propagate a large file over a gossip network.
     ///
     /// <https://github.com/tendermint/tendermint/wiki/Block-Structure#partset>
-    pub parts: Option<parts::Header>,
+    pub parts: Option<PartSetHeader>,
+}
+
+impl DomainType<RawBlockId> for Id {}
+
+impl TryFrom<RawBlockId> for Id {
+    type Error = Error;
+
+    fn try_from(value: RawBlockId) -> Result<Self, Self::Error> {
+        Ok(Self {
+            hash: value.hash.try_into()?,
+            parts: match value.part_set_header {
+                None => None,
+                Some(p) => Some(p.try_into()?),
+            },
+        })
+    }
+}
+
+impl From<Id> for RawBlockId {
+    fn from(value: Id) -> Self {
+        // https://github.com/tendermint/tendermint/blob/1635d1339c73ae6a82e062cd2dc7191b029efa14/types/block.go#L1204
+        // The Go implementation encodes a nil value into an empty struct. We try our best to
+        // anticipate an empty struct by using the default implementation which would otherwise be
+        // invalid.
+        if value == Id::default() {
+            RawBlockId {
+                hash: vec![],
+                part_set_header: Some(RawPartSetHeader {
+                    total: 0,
+                    hash: vec![],
+                }),
+            }
+        } else {
+            RawBlockId {
+                hash: value.hash.into(),
+                part_set_header: value.parts.map(|p| p.into()),
+            }
+        }
+    }
+}
+
+impl TryFrom<RawCanonicalBlockId> for Id {
+    type Error = Error;
+
+    fn try_from(value: RawCanonicalBlockId) -> Result<Self, Self::Error> {
+        Ok(Self {
+            hash: value.hash.try_into()?,
+            parts: match value.part_set_header {
+                None => None,
+                Some(p) => Some(p.try_into()?),
+            },
+        })
+    }
+}
+
+impl From<Id> for RawCanonicalBlockId {
+    // Todo: it's possible that an empty CanonicalBlockId is encoded differently. Test this.
```

Comment:
Just making sure we do not forget about this TODO.

Reply:
I just looked into it and rearranged the BlockID to reflect better on the Go implementation and the differences between the protobuf message and the Go expectations. You will see it in the next commit, with comments.